File: sim/core/bulk/batch.go
package bulk

import (
	"sync"

	"github.com/wowsims/tbc/sim/core"
	"github.com/wowsims/tbc/sim/core/proto"
	"github.com/wowsims/tbc/sim/core/simsignals"
	googleProto "google.golang.org/protobuf/proto"
)

type BulkSimStageTask struct {
	Candidate BulkSimCandidate
	Position  int
}

// Describes one pass of candidate sims: how much to run, where its progress sits
// within the stage, and which already-simmed metrics to fold into each result.
type bulkSimCandidateBatchConfig struct {
	iterations              int32
	seedOffset              int32
	concurrency             int
	completedSimsBase       int
	completedIterationsBase int32
	emitter                 bulkSimStageProgressEmitter
	carried                 *bulkSimStageCarryOver
}

// Runs `iterations` for every candidate across `concurrency` workers, folding in any
// carried-over metrics, and returns the collected results. Both the initial stage pass
// and the adaptive extra-iteration passes go through here so their queueing, progress
// accounting, abort handling and merging cannot drift apart.
func runBulkSimCandidateBatch(request *proto.BulkSimRequest, candidates []BulkSimCandidate, config BulkSimStageConfig, batch bulkSimCandidateBatchConfig, signals simsignals.Signals) []*BulkSimCandidateResult {
	jobs := make(chan BulkSimStageTask, len(candidates))
	// Results are written by position rather than collected from a channel: the slots
	// are disjoint, so no synchronisation is needed and the output order is the
	// candidate order instead of the completion order.
	resultsByPosition := make([]*BulkSimCandidateResult, len(candidates))
	progressTracker := &BulkSimStageProgressTracker{
		emitter:                             batch.emitter,
		totalCandidates:                     len(candidates),
		iterations:                          batch.iterations,
		completedSimsBeforeCandidates:       batch.completedSimsBase,
		completedIterationsBeforeCandidates: batch.completedIterationsBase,
		completedIterationsByCandidate:      make([]int32, len(candidates)),
	}
	var wg sync.WaitGroup

	for range max(1, batch.concurrency) {
		wg.Go(func() {
			for task := range jobs {
				if signals.Abort.IsTriggered() {
					return
				}

				candidateResult := runSingleBulkSimCandidate(request, task.Candidate, batch.iterations, batch.seedOffset, signals, config.UseConcurrentSim, func(progressMetrics *proto.ProgressMetrics) {
					progressTracker.reportCandidateProgress(task.Position, progressMetrics)
				})
				candidateResult = mergeBulkSimCandidateResults(batch.carried.candidateResult(task.Candidate.Index), candidateResult)
				progressTracker.reportCandidateComplete(task.Position, candidateResult)
				resultsByPosition[task.Position] = candidateResult
				if candidateResult.Error != nil {
					signals.Abort.Trigger()
				}
			}
		})
	}

	go func() {
		defer close(jobs)
		for idx, candidate := range candidates {
			if signals.Abort.IsTriggered() {
				return
			}
			jobs <- BulkSimStageTask{Candidate: candidate, Position: idx}
		}
	}()
	wg.Wait()

	// Aborted runs leave gaps where a candidate never started.
	return core.FilterSlice(resultsByPosition, func(candidateResult *BulkSimCandidateResult) bool {
		return candidateResult != nil
	})
}

// Runs one gear set start to finish with no progress reporting, for the paths that just
// need a number (the baseline of a run with no candidates).
func runSingleBulkSim(request *proto.BulkSimRequest, candidate BulkSimCandidate, iterations int32, signals simsignals.Signals) *BulkSimCandidateResult {
	return runSingleBulkSimCandidate(request, candidate, iterations, 0, signals, false, nil)
}

// Runs `iterations` for one gear set, starting the RNG `seedOffset` iterations into the
// sequence so a segment can extend an earlier run instead of repeating it.
func runSingleBulkSimCandidate(request *proto.BulkSimRequest, candidate BulkSimCandidate, iterations int32, seedOffset int32, signals simsignals.Signals, useConcurrentSim bool, progressCallback func(*proto.ProgressMetrics)) *BulkSimCandidateResult {
	if signals.Abort.IsTriggered() {
		return &BulkSimCandidateResult{Candidate: candidate, Error: bulkSimAbortedError()}
	}

	simRequest := googleProto.Clone(request.BaseRequest).(*proto.RaidSimRequest)
	simRequest.SimOptions.Iterations = iterations
	simRequest.SimOptions.RandomSeed += int64(seedOffset)
	simRequest.SimOptions.DebugFirstIteration = false
	simRequest.SimOptions.Debug = false
	// Every candidate runs the same seed sequence, so keeping the per-iteration values
	// lets culling compare paired differences instead of marginal errors.
	simRequest.SimOptions.SaveAllValues = true

	player, err := getBulkSimPlayer(simRequest.Raid)
	if err != "" {
		return &BulkSimCandidateResult{Candidate: candidate, Error: &proto.ErrorOutcome{Message: err}}
	}
	player.Equipment = googleProto.Clone(candidate.Gear).(*proto.EquipmentSpec)
	adjustCandidateImbues(player)

	var simProgress chan *proto.ProgressMetrics
	var progressWg sync.WaitGroup
	if progressCallback != nil && !simRequest.SimOptions.IsTest {
		simProgress = make(chan *proto.ProgressMetrics, 16)
		progressWg.Go(func() {
			for progressMetrics := range simProgress {
				progressCallback(progressMetrics)
			}
		})
	}

	var simResult *proto.RaidSimResult
	if useConcurrentSim && progressCallback == nil {
		simResult = core.RunRaidSimConcurrentWithSignals(simRequest, signals)
	} else {
		simResult = core.RunSim(simRequest, simProgress, signals)
	}
	if simProgress != nil {
		progressWg.Wait()
	}
	if simResult == nil {
		return &BulkSimCandidateResult{Candidate: candidate, Error: &proto.ErrorOutcome{Message: "Bulk sim did not return a result"}}
	}
	if simResult.Error != nil {
		if simResult.Error.Type == proto.ErrorOutcomeType_ErrorOutcomeAborted && simResult.Error.Message == "" {
			return &BulkSimCandidateResult{Candidate: candidate, Error: bulkSimAbortedError()}
		}
		return &BulkSimCandidateResult{Candidate: candidate, Error: simResult.Error}
	}

	return &BulkSimCandidateResult{
		Candidate:  candidate,
		DpsMetrics: cleanBulkSimDpsMetrics(simResult.RaidMetrics.Dps),
	}
}

File: ui/core/wasm/bulk_sim/batch.ts
import { queue } from 'async';
import { BulkSimRequest, ErrorOutcome, ErrorOutcomeType, ProgressMetrics, RaidSimRequest, RaidSimResult } from '../../proto/api';
import { Database } from '../../proto_utils/database';
import { SimSignals } from '../../sim_signal_manager';
import { noop } from '../../utils';
import { WorkerPool, WorkerProgressCallback } from '../../worker_pool';
import { runConcurrentSim } from '../sim';
import { cleanBulkSimDpsMetrics, mergeBulkSimCandidateResults } from './merge';
import { BulkSimStageProgressEmitter } from './progress';
import { ConcurrentBulkSimCandidate, ConcurrentBulkSimCandidateResult, ConcurrentBulkSimCandidateTask, ConcurrentBulkSimStageConfig } from './types';

export type ConcurrentBulkSimCandidateBatchConfig = {
	completedSimsBase: number;
	completedIterationsBase: number;
	emitter: BulkSimStageProgressEmitter;
	seedOffset?: number;
};

const makeBulkSimRequestForCandidate = (request: BulkSimRequest, candidate: ConcurrentBulkSimCandidate, iterations: number, seedOffset = 0): RaidSimRequest => {
	const simRequest = RaidSimRequest.clone(request.baseRequest!);
	simRequest.requestId = request.requestId;
	simRequest.simOptions!.iterations = iterations;
	simRequest.simOptions!.randomSeed += BigInt(seedOffset);
	simRequest.simOptions!.debugFirstIteration = false;
	simRequest.simOptions!.debug = false;
	// Every candidate runs the same seed sequence, so keeping the per-iteration values
	// lets culling compare paired differences instead of marginal errors.
	simRequest.simOptions!.saveAllValues = true;
	const player = simRequest.raid!.parties[0].players[0];
	player.equipment = candidate.gear;
	// Keep weapon stone imbues in sync with this candidate's weapon types, mirroring the
	// frontend auto-switch so bulk combos use the correct stone (or none).
	if (player.consumables && candidate.gear) {
		player.consumables = Database.getSync().lookupEquipmentSpec(candidate.gear).adjustImbues(player.consumables);
	}
	return simRequest;
};

export enum BulkSimCandidateTransport {
	// Spreads one gear set's iterations across every worker. Used for the baseline,
	// which has no siblings to run alongside it.
	SplitAcrossWorkers = 'SplitAcrossWorkers',
	// Keeps a gear set on a single worker. Used by the candidate batch, which saturates
	// the pool with candidates instead.
	SingleWorker = 'SingleWorker',
}

export const runSingleBulkSimCandidate = async (
	request: BulkSimRequest,
	candidate: ConcurrentBulkSimCandidate,
	iterations: number,
	workerPool: WorkerPool,
	signals: SimSignals,
	transport: BulkSimCandidateTransport,
	progressCallback?: (progressMetrics: ProgressMetrics) => void,
	seedOffset = 0,
): Promise<ConcurrentBulkSimCandidateResult> => {
	if (signals.abort.isTriggered()) {
		return { candidate, error: ErrorOutcome.create({ type: ErrorOutcomeType.ErrorOutcomeAborted }) };
	}

	const simRequest = makeBulkSimRequestForCandidate(request, candidate, iterations, seedOffset);
	let simResult: RaidSimResult;
	if (transport === BulkSimCandidateTransport.SplitAcrossWorkers) {
		simResult = await runConcurrentSim(simRequest, workerPool, progressCallback ?? noop, signals);
	} else {
		simRequest.requestId = `${request.requestId}-${candidate.index}-${seedOffset}`;
		simResult = await workerPool.raidSimAsync(simRequest, progressCallback ?? noop, signals);
	}
	if (simResult.error) {
		return { candidate, error: simResult.error };
	}

	return {
		candidate,
		dpsMetrics: cleanBulkSimDpsMetrics(simResult.raidMetrics?.dps),
	};
};

export const runBulkSimCandidateBatchOnWorkers = async (
	request: BulkSimRequest,
	candidates: ConcurrentBulkSimCandidate[],
	iterations: number,
	workerPool: WorkerPool,
	signals: SimSignals,
	batchConfig: ConcurrentBulkSimCandidateBatchConfig,
	carriedByCandidate?: Map<number, ConcurrentBulkSimCandidateResult>,
): Promise<ConcurrentBulkSimCandidateResult[]> => {
	const candidateIterationsDone = Array(candidates.length).fill(0);
	const results: Array<ConcurrentBulkSimCandidateResult | undefined> = [];
	const concurrency = Math.max(1, Math.min(workerPool.getNumWorkers(), candidates.length));
	let completedCandidates = 0;
	let completedCandidateIterations = 0;

	const updateCandidateIterations = (idx: number, completedIterations: number) => {
		const nextCompletedIterations = Math.min(completedIterations, iterations);
		completedCandidateIterations += nextCompletedIterations - candidateIterationsDone[idx];
		candidateIterationsDone[idx] = nextCompletedIterations;
	};

	const candidateQueue = queue<ConcurrentBulkSimCandidateTask, Error>(async ({ candidate, idx }) => {
		if (signals.abort.isTriggered()) return;

		const simmed = await runSingleBulkSimCandidate(
			request,
			candidate,
			iterations,
			workerPool,
			signals,
			BulkSimCandidateTransport.SingleWorker,
			progressMetrics => {
				if (progressMetrics.totalIterations == 0) return;
				updateCandidateIterations(idx, progressMetrics.completedIterations);
				batchConfig.emitter.report(
					batchConfig.completedSimsBase + completedCandidates,
					batchConfig.completedIterationsBase + completedCandidateIterations,
					progressMetrics.dps,
				);
			},
			batchConfig.seedOffset,
		);

		const candidateResult = mergeBulkSimCandidateResults(carriedByCandidate?.get(candidate.index), simmed);
		updateCandidateIterations(idx, iterations);
		completedCandidates++;
		batchConfig.emitter.report(
			batchConfig.completedSimsBase + completedCandidates,
			batchConfig.completedIterationsBase + completedCandidateIterations,
			candidateResult.dpsMetrics?.avg ?? 0,
		);

		if (candidateResult.error) {
			signals.abort.trigger();
		}
		results[idx] = candidateResult;
	}, concurrency);

	const queueErrorPromise = candidateQueue.error();
	candidates.forEach((candidate, idx) => candidateQueue.push({ candidate, idx }));
	await Promise.race([candidateQueue.drain(), queueErrorPromise]);
	return results.filter((result): result is ConcurrentBulkSimCandidateResult => !!result);
};
